crates/ty_python_semantic/resources/mdtest/public_types.md

```diff
@@ -0,0 +1,421 @@
+
+```py
+class A: ...
+
+def outer(x: A | None):
+    if x is not None:
+        def inner() -> None:
+            # TODO: should ideally be `A`
+            reveal_type(x)  # revealed: A | None
+        inner()
+```
+
+### Shadowing
+
+Similarly, since we do not analyze control flow in the outer scope here, we assume that `inner()`
+could be called between the two assignments to `x`:
+
+```py
+def outer() -> None:
+    def inner() -> None:
+        # TODO: this should ideally be `Unknown | Literal[1]`, but no other type checker supports this either
+        reveal_type(x)  # revealed: Unknown | None | Literal[1]
+    x = None
+
+    # [additional code here]
+
+    x = 1
+
+    inner()
+```
+
+This is currently even true if the `inner` function is only defined after the second assignment to
+`x`:
+
+```py
+def outer() -> None:
+    x = None
+
+    # [additional code here]
+
+    x = 1
+
+    def inner() -> None:
+        # TODO: this should be `Unknown | Literal[1]`. Mypy and pyright support this.
+        reveal_type(x)  # revealed: Unknown | None | Literal[1]
+    inner()
+```
+
+A similar case derived from an ecosystem example, involving declared types:
+
+```py
+class C: ...
+
+def outer(x: C | None):
+    x = x or C()
+
+    reveal_type(x)  # revealed: C
+
+    def inner() -> None:
+        # TODO: this should ideally be `C`
+        reveal_type(x)  # revealed: C | None
+    inner()
+```
+
+### Assignments to nonlocal variables
+
+Writes to the outer-scope variable are currently not detected:
+
+```py
+def outer() -> None:
+    x = None
+
+    def set_x() -> None:
+        nonlocal x
+        x = 1
+    set_x()
+
+    def inner() -> None:
+        # TODO: this should ideally be `Unknown | None | Literal[1]`.
+        reveal_type(x)  # revealed: Unknown | None
+    inner()
+```
+
+## Handling of overloads
+
+### With implementation
+
+Overloads need special treatment, because here, we do not want to consider *all* possible
+definitions of `f`. This would otherwise result in a union of all three definitions of `f`:
+
+```py
+from typing import overload
+
+@overload
+def f(x: int) -> int: ...
+@overload
+def f(x: str) -> str: ...
+def f(x: int | str) -> int | str:
+    raise NotImplementedError
+
+reveal_type(f)  # revealed: Overload[(x: int) -> int, (x: str) -> str]
+
+def _():
+    reveal_type(f)  # revealed: Overload[(x: int) -> int, (x: str) -> str]
+```
+
+This also works if there are conflicting declarations:
+
+```py
+def flag() -> bool:
+    return True
+
+if flag():
+    @overload
+    def g(x: int) -> int: ...
+    @overload
+    def g(x: str) -> str: ...
+    def g(x: int | str) -> int | str:
+        return x
+
+else:
+    g: str = ""
+
+def _():
+    reveal_type(g)  # revealed: (Overload[(x: int) -> int, (x: str) -> str]) | str
+
+# error: [conflicting-declarations]
+g = "test"
+```
+
+### Without an implementation
+
+If there is no implementation, we only consider the last overload definition.
```

Comment:
I'm struggling to parse this -- the example seems to show that we include both overloads in the revealed type of `f`, just like we do if there _is_ an overload implementation?